File: examples/practical-app/tests/e2e/todo.spec.ts
import { test, expect } from './fixtures';

test.describe('Todo Functionality', () => {
  test.beforeEach(async ({ page }) => {
    await page.goto('http://localhost:3001/');
    await page.waitForLoadState('networkidle');
    // Clear localStorage to ensure clean state for each test
    await page.evaluate(() => localStorage.clear());
    await page.reload();
    await page.waitForLoadState('networkidle');
  });

  test('should add a new todo', async ({ page }) => {
    const input = page.locator('[data-todo-target="input"]');
    const todoList = page.locator('[data-todo-target="list"]');

    // Add a new todo
    await input.fill('Buy groceries');
    await page.keyboard.press('Enter');

    // Verify todo was added
    await expect(todoList.locator('text=Buy groceries')).toBeVisible();

    // Input should be cleared
    await expect(input).toHaveValue('');
  });

  test('should toggle todo completion', async ({ page }) => {
    // Add a todo first
    await page.locator('[data-todo-target="input"]').fill('Test todo');
    await page.keyboard.press('Enter');

    // Find the checkbox and todo item
    const checkbox = page.locator('[data-todo-target="list"] input[type="checkbox"]').first();
    const todoItem = page.locator('[data-todo-target="list"] .todo-item').first();

    // Initially unchecked
    await expect(checkbox).not.toBeChecked();
    await expect(todoItem).not.toHaveClass(/completed/);

    // Toggle completion
    await checkbox.click();

    // Should be checked and styled (completed class adds line-through via CSS)
    await expect(checkbox).toBeChecked();
    await expect(todoItem).toHaveClass(/completed/);
  });

  test('should delete a todo', async ({ page }) => {
    // Add a todo
    await page.locator('[data-todo-target="input"]').fill('Todo to delete');
    await page.keyboard.press('Enter');

    // Verify it exists
    await expect(page.locator('text=Todo to delete')).toBeVisible();

    // Click delete button
    const deleteButton = page.locator('[data-todo-target="list"] button[data-action*="delete"]').first();
    await deleteButton.click();

    // Todo should be removed
    await expect(page.locator('text=Todo to delete')).not.toBeVisible();
  });

  test('should filter completed todos', async ({ page }) => {
    // Add multiple todos
    await page.locator('[data-todo-target="input"]').fill('Todo 1');
    await page.keyboard.press('Enter');

    await page.locator('[data-todo-target="input"]').fill('Todo 2');
    await page.keyboard.press('Enter');

    // Complete first todo
    await page.locator('[data-todo-target="list"] input[type="checkbox"]').first().click();

    // Click "Completed" filter button (using specific selector to avoid matching count text)
    await page.click('button[data-filter="completed"]');

    // Should show only completed todo
    await expect(page.locator('[data-todo-target="list"]').locator('text=Todo 1')).toBeVisible();
    await expect(page.locator('[data-todo-target="list"]').locator('text=Todo 2')).not.toBeVisible();

    // Click "Active" filter button
    await page.click('button[data-filter="active"]');

    // Should show only active todo
    await expect(page.locator('[data-todo-target="list"]').locator('text=Todo 1')).not.toBeVisible();
    await expect(page.locator('[data-todo-target="list"]').locator('text=Todo 2')).toBeVisible();

    // Click "All" filter button
    await page.click('button[data-filter="all"]');

    // Should show both
    await expect(page.locator('[data-todo-target="list"]').locator('text=Todo 1')).toBeVisible();
    await expect(page.locator('[data-todo-target="list"]').locator('text=Todo 2')).toBeVisible();
  });

  test('should persist todos in localStorage', async ({ page }) => {
    // Add a todo
    await page.locator('[data-todo-target="input"]').fill('Persistent todo');
    await page.keyboard.press('Enter');

    // Reload page
    await page.reload();
    await page.waitForLoadState('networkidle');

    // Todo should still be there
    await expect(page.locator('text=Persistent todo')).toBeVisible();
  });

  test('should show todo count', async ({ page }) => {
    // Check initial count
    const countElement = page.locator('[data-todo-target="count"]');

    // Add todos
    await page.locator('[data-todo-target="input"]').fill('Todo 1');
    await page.keyboard.press('Enter');

    await page.locator('[data-todo-target="input"]').fill('Todo 2');
    await page.keyboard.press('Enter');

    // Count should update
    await expect(countElement).toContainText('2');

    // Complete one todo
    await page.locator('[data-todo-target="list"] input[type="checkbox"]').first().click();

    // Active count should decrease
    await expect(countElement).toContainText('1');
  });

  test('should not add empty todos', async ({ page }) => {
    const input = page.locator('[data-todo-target="input"]');
    const todoList = page.locator('[data-todo-target="list"]');

    // Get initial todo count
    const initialCount = await todoList.locator('.todo-item').count();

    // Try to add empty todo
    await input.fill('   '); // Just spaces
    await page.keyboard.press('Enter');

    // Count should not change
    const finalCount = await todoList.locator('.todo-item').count();
    expect(finalCount).toBe(initialCount);
  });

  test('should clear completed todos', async ({ page }) => {
    // Add and complete multiple todos
    for (let i = 1; i <= 3; i++) {
      await page.locator('[data-todo-target="input"]').fill(`Todo ${i}`);
      await page.keyboard.press('Enter');
    }

    // Wait for all todos to be added
    await expect(page.locator('[data-todo-target="list"] .todo-item')).toHaveCount(3);

    // Complete first two
    const checkboxes = page.locator('[data-todo-target="list"] input[type="checkbox"]');
    await checkboxes.nth(0).click();
    await checkboxes.nth(1).click();

    // Click "Clear Completed" button (using specific selector)
    await page.click('button:has-text("Clear Completed")');

    // Only active todo should remain in the list
    await expect(page.locator('[data-todo-target="list"]').locator('text=Todo 1')).not.toBeVisible();
    await expect(page.locator('[data-todo-target="list"]').locator('text=Todo 2')).not.toBeVisible();
    await expect(page.locator('[data-todo-target="list"]').locator('text=Todo 3')).toBeVisible();
  });
});

File: examples/practical-app/spec/features/todo_spec.rb
# frozen_string_literal: true

RSpec.describe 'Todo Functionality', type: :feature do
  let(:input_selector) { '[data-todo-target="input"]' }
  let(:list_selector) { '[data-todo-target="list"]' }

  # Add todo using stable helpers with explicit count verification
  def add_todo(text)
    # Get current count before adding
    current_count = all("#{list_selector} .todo-item", wait: 0).count

    stable_input(input_selector, text, submit_key: :enter)

    # Wait for the new todo item to appear (count should increase by 1)
    wait_for_count("#{list_selector} .todo-item", current_count + 1)
    wait_for_text(list_selector, text)
    wait_for_dom_stable
  end

  def fill_in_todo_input(text)
    stable_set(input_selector, text)
  end

  describe 'adding todos' do
    it 'adds a new todo' do
      add_todo('Buy groceries')

      # Wait for the todo to appear in the list
      expect(page).to have_css("#{list_selector} .todo-item", wait: 5)

      within(list_selector) do
        expect(page).to have_content('Buy groceries')
      end

      # Input should be cleared
      expect(find(input_selector).value).to eq('')
    end

    it 'does not add empty todos' do
      initial_count = all("#{list_selector} .todo-item").count

      fill_in_todo_input('   ')
      find(input_selector).native.send_keys(:enter)

      expect(all("#{list_selector} .todo-item").count).to eq(initial_count)
    end
  end

  describe 'toggling todos' do
    before do
      add_todo('Test todo')
      expect(page).to have_css("#{list_selector} .todo-item", count: 1, wait: 5)
    end

    it 'toggles todo completion' do
      checkbox = find("#{list_selector} input[type='checkbox']", match: :first, wait: 5)
      todo_item = find("#{list_selector} .todo-item", match: :first)

      # Initially unchecked
      expect(checkbox).not_to be_checked
      expect(todo_item[:class]).not_to include('completed')

      # Toggle completion
      checkbox.click

      # Should be checked and have completed class
      expect(checkbox).to be_checked
      expect(todo_item[:class]).to include('completed')
    end
  end

  describe 'deleting todos' do
    it 'deletes a todo' do
      add_todo('Todo to delete')

      # Wait for the todo item to appear
      expect(page).to have_css("#{list_selector} .todo-item", wait: 5)
      expect(page).to have_content('Todo to delete')

      # Click delete button
      find("#{list_selector} button[data-action*='delete']", match: :first).click

      # Wait for deletion to complete
      expect(page).not_to have_css("#{list_selector} .todo-item", wait: 5)
      expect(page).not_to have_content('Todo to delete')
    end
  end

  describe 'filtering todos' do
    before do
      add_todo('Todo 1')
      expect(page).to have_css("#{list_selector} .todo-item", count: 1, wait: 5)

      add_todo('Todo 2')
      expect(page).to have_css("#{list_selector} .todo-item", count: 2, wait: 5)

      # Complete first todo
      find("#{list_selector} input[type='checkbox']", match: :first).click
      expect(page).to have_css("#{list_selector} .todo-item.completed", count: 1, wait: 5)
    end

    it 'filters completed todos' do
      click_button_with_filter('completed')

      within(list_selector) do
        expect(page).to have_content('Todo 1')
        expect(page).not_to have_content('Todo 2')
      end
    end

    it 'filters active todos' do
      click_button_with_filter('active')

      within(list_selector) do
        expect(page).not_to have_content('Todo 1')
        expect(page).to have_content('Todo 2')
      end
    end

    it 'shows all todos' do
      click_button_with_filter('completed')
      click_button_with_filter('all')

      within(list_selector) do
        expect(page).to have_content('Todo 1')
        expect(page).to have_content('Todo 2')
      end
    end

    def click_button_with_filter(filter)
      find("button[data-filter='#{filter}']").click
    end
  end

  describe 'persistence' do
    it 'persists todos in localStorage' do
      add_todo('Persistent todo')
      # Wait for todo to be added to DOM and localStorage
      expect(page).to have_css("#{list_selector} .todo-item", count: 1, wait: 5)

      # Reload page
      visit '/'
      expect(page).to have_css('[data-controller]', wait: 10)

      expect(page).to have_content('Persistent todo', wait: 5)
    end
  end

  describe 'todo count' do
    let(:count_selector) { '[data-todo-target="count"]' }

    it 'shows todo count' do
      add_todo('Todo 1')
      expect(page).to have_css("#{list_selector} .todo-item", count: 1, wait: 5)

      add_todo('Todo 2')
      expect(page).to have_css("#{list_selector} .todo-item", count: 2, wait: 5)

      # Wait for count to update
      expect(page).to have_css(count_selector, text: /2/, wait: 5)

      # Complete one todo
      find("#{list_selector} input[type='checkbox']", match: :first).click

      # Wait for count to update after completion
      expect(page).to have_css(count_selector, text: /1/, wait: 5)
    end
  end

  describe 'clearing completed' do
    it 'clears completed todos' do
      # Add todos one by one with waits
      add_todo("Todo 1")
      expect(page).to have_css("#{list_selector} .todo-item", count: 1, wait: 5)

      add_todo("Todo 2")
      expect(page).to have_css("#{list_selector} .todo-item", count: 2, wait: 5)

      add_todo("Todo 3")
      expect(page).to have_css("#{list_selector} .todo-item", count: 3, wait: 5)

      # Complete first checkbox - get fresh reference
      all("#{list_selector} input[type='checkbox']")[0].click
      # Wait for completion state to be applied
      expect(page).to have_css("#{list_selector} .todo-item.completed", count: 1, wait: 5)

      # Complete second checkbox - get fresh reference again
      all("#{list_selector} input[type='checkbox']")[1].click
      # Wait for completion state
      expect(page).to have_css("#{list_selector} .todo-item.completed", count: 2, wait: 5)

      # Click "Clear Completed" button
      click_button 'Clear Completed'

      # Wait for clearing to complete - should have 1 remaining
      expect(page).to have_css("#{list_selector} .todo-item", count: 1, wait: 5)

      within(list_selector) do
        expect(page).not_to have_content('Todo 1')
        expect(page).not_to have_content('Todo 2')
        expect(page).to have_content('Todo 3')
      end
    end
  end
end
